File: Day 2/solution.py
"""
Day 2 solution.
"""
import sys
from typing import Generator


class Solver:
    """Day 2 solver."""

    def __init__(self, filepath: str = "input.txt"):
        self.filepath = filepath

    def decode_line(self, line: str) -> tuple[int, list[str]]:
        """
        Decode line to get game id and draw details for the game.
        """
        game_detail, draw_details = line.split(":")
        return (int(game_detail.split()[-1]), draw_details.split(";"))

    def decode_draw(
        self, game_set: str
    ) -> Generator[tuple[int, int], None, None]:
        """
        Decodes set details to a list of colour id, number of colour pairs.
        """
        for ball_details in game_set.split(", "):
            num_token, colour_token = ball_details.split()
            yield int(
                num_token
            ), 0 if colour_token == "red" else 1 if colour_token == "green" else 2

    def part_1(self):
        """Part 1 solver."""
        max_colours = [12, 13, 14]

        def all_draws_less_than_max(draws: list[str]) -> bool:
            def draw_less_than_max(draw: str) -> bool:
                return all(
                    num <= max_colours[colour]
                    for num, colour in self.decode_draw(draw)
                )

            return all(draw_less_than_max(draw) for draw in draws)

        with open(
            self.filepath, "r", encoding=sys.getdefaultencoding()
        ) as file:
            return sum(
                game_id
                for game_id, draws in (self.decode_line(line) for line in file)
                if all_draws_less_than_max(draws)
            )

    def part_2(self):
        """Part 2 solver."""

        def get_min_colours(draws: list[str]) -> list[int]:
            colours = [0, 0, 0]
            for draw in draws:
                for num, colour in self.decode_draw(draw):
                    colours[colour] = max(colours[colour], num)
            return colours

        def multiply(colours: list[int]) -> int:
            result = 1
            for x in colours:
                result *= x
            return result

        with open(
            self.filepath, "r", encoding=sys.getdefaultencoding()
        ) as file:
            return sum(
                multiply(get_min_colours(draws))
                for _, draws in (self.decode_line(line) for line in file)
            )

    def solve(self) -> None:
        """Runs part 1 and part 2."""
        print("Day 2")
        print(f"Solving: {self.filepath}")
        print("Part 1:")
        print(self.part_1())
        print("---")
        print("Part 2:")
        print(self.part_2())


if __name__ == "__main__":
    match sys.argv:
        case [_]:
            solver = Solver()
        case [_, "--path", path] | [_, "-p", path] | [_, path]:
            solver = Solver(path)
        case other:
            print("Usage: python solution.py <path>")
            sys.exit(0)
    solver.solve()

File: Day 2/solution.cpp
#include <filesystem>
#include <fstream>
#include <iostream>
#include <vector>
#include <cassert>

struct FileParserIterator
{
	using iterator_category = std::input_iterator_tag;
	using value_type = std::pair<int, std::vector<std::string>>;
	using reference = value_type const &;
	using pointer = value_type const *;
	using difference_type = ptrdiff_t;

	FileParserIterator(std::filesystem::path filepath, bool done = false) : done(done)
	{
		this->file = new std::ifstream();
		this->file->open(filepath);
		done || ++*this;
	}

	FileParserIterator(const FileParserIterator &other) : done(other.done),
																												value(other.value), file(other.file) {}

	~FileParserIterator()
	{
		this->file->close();
		delete this->file;
	}

	explicit operator bool() const { return !done; }

	reference operator*() const { return value; }
	pointer operator->() const { return &value; }

	FileParserIterator &operator++()
	{
		assert(!this->done);

		this->done = this->file->eof();
		if (this->done)
			return *this;

		int game_id = 0;
		std::vector<std::string> sets;

		// Read a line
		char c = 0;
		while (!this->file->eof() && c != '\n')
		{
			while ((c = this->file->get()) != ':')
			{
				if (!std::isdigit(c))
					continue;
				game_id = game_id * 10 + (c - '0');
			}

			while (!this->file->eof() && c != '\n')
			{
				std::string current;
				while ((c = this->file->get()) != ';' && c != '\n' && c != EOF)
				{
					current.push_back(c);
				}
				sets.push_back(current);
				current.clear();
			}
		}

		this->value = std::make_pair(game_id, sets);
		return *this;
	}

	FileParserIterator operator++(int)
	{
		FileParserIterator const tmp(*this);
		++*this;
		return tmp;
	}

	friend bool operator!=(const FileParserIterator &a, const FileParserIterator &b)
	{
		return a.done != b.done;
	};

private:
	bool done;
	value_type value;
	std::ifstream *file = nullptr;
};

class FileParser
{
	std::filesystem::path filepath;
	using Iterator = FileParserIterator;

public:
	FileParser(std::filesystem::path filepath) : filepath(filepath) {}

	Iterator begin() const
	{
		return Iterator(this->filepath);
	}

	Iterator end() const
	{
		return Iterator(this->filepath, true);
	}
};

class Solver
{
	std::filesystem::path filepath;

	std::vector<std::pair<int, int>> decode_set(const std::string &game_set)
	{
		std::vector<std::pair<int, int>> result;
		for (int i = 0; i < game_set.size(); ++i)
		{
			int num = 0;
			std::string colour;

			for (; i < game_set.size() && game_set[i] != ','; ++i)
			{
				char c = game_set[i];
				if (std::isdigit(c))
				{
					num = num * 10 + (c - '0');
				}
				else if (c != ' ')
				{
					colour += c;
				}
			}

			result.emplace_back(num, colour == "red" ? 0 : (colour == "green" ? 1 : 2));
		}
		return result;
	}

	int part_1()
	{
		FileParser file(this->filepath);
		int result = 0, max_colours[3]{12, 13, 14};

		for (auto [game_id, sets] : file)
		{
			bool valid = true;
			for (const std::string &set : sets)
			{
				if (!valid)
					break;
				for (const auto [num, colour] : this->decode_set(set))
				{
					if (num > max_colours[colour])
					{
						valid = false;
						break;
					}
				}
			}
			if (valid)
			{
				result += game_id;
			}
		}
		return result;
	}

	int part_2()
	{
		FileParser file(this->filepath);
		int result = 0;

		for (auto [_, sets] : file)
		{
			int colours[3]{0, 0, 0};
			for (const std::string &set : sets)
			{
				for (const auto [num, colour] : this->decode_set(set))
				{
					colours[colour] = std::max(colours[colour], num);
				}
			}

			result += colours[0] * colours[1] * colours[2];
		}
		return result;
	}

public:
	Solver(std::filesystem::path filepath = "input.txt") : filepath(filepath) {}

	void solve()
	{
		std::cout << "Day 1" << std::endl;
		std::cout << "Solving: " << this->filepath << std::endl;
		std::cout << "Part 1:" << std::endl;
		std::cout << this->part_1() << std::endl;
		std::cout << "---" << std::endl;
		std::cout << "Part 2:" << std::endl;
		std::cout << this->part_2() << std::endl;
	}
};

int main(int argc, char *argv[])
{
	std::unique_ptr<Solver> solver;
	if (argc == 1)
	{
		solver = std::make_unique<Solver>();
	}
	else if (argc == 2)
	{
		solver = std::make_unique<Solver>(argv[1]);
	}
	else if (argc == 3 && (argv[1] == "-p" || argv[1] == "--path"))
	{
		solver = std::make_unique<Solver>(argv[2]);
	}
	else
	{
		return 1;
	}

	solver->solve();
}